packages/ui/feature-builder-right-sidebar/src/lib/edit-step-sidebar/edit-step-form-container/edit-step-form-container.component.ts

```diff
@@ -96,55 +99,74 @@ export class EditStepFormContainerComponent {
-          .pipe(take(1));
-      }),
-      switchMap((res) => {
-        if (res?.type === TriggerType.PIECE) {
-          return this.actionMetaService
-            .getPieceMetadata(res.settings.pieceName, res.settings.pieceVersion)
-            .pipe(
-              map((meta) => {
-                return { step: res, metadata: meta };
-              })
-            );
-        }
-        return of({ step: res, metadata: undefined });
-      }),
-      tap((res) => {
-        if (
-          this._selectedStep.type === TriggerType.PIECE ||
-          this._selectedStep.type === TriggerType.WEBHOOK
-        ) {
-          if (this._selectedStep.type === TriggerType.WEBHOOK) {
-            this.updateNonAppWebhookTrigger(res.step!);
-          } else {
-            const newTriggerSettings = this.createPieceSettings(res.step!);
-            const trigger =
-              res.metadata?.triggers[newTriggerSettings.triggerName];
-            if (trigger?.type === TriggerStrategy.APP_WEBHOOK) {
-              this.updateAppWebhookTrigger(res.step!, trigger);
-            } else {
-              this.updateNonAppWebhookTrigger(res.step!);
-            }
-          }
-        } else {
-          this.store.dispatch(
-            FlowsActions.updateAction({
-              operation: this.prepareStepDataToSave(
-                res.step!
-              ) as UpdateActionRequest,
-            })
-          );
-        }
-      }),
-      map(() => void 0)
+    const delayForWriteValueToTakeEffectThenListenToSaving = of(null).pipe(
+      delay(10)
     );
+    this.builderService.savingStepOrTriggerData$ =
+      delayForWriteValueToTakeEffectThenListenToSaving.pipe(
+        switchMap(() => {
+          return this.stepForm.valueChanges.pipe(
+            takeUntil(this.cancelAutoSaveListener$),
+            skipWhile(() => this.stepForm.disabled),
+            delay(1), //to ensure name doesn't get overwritten from changing selected action/trigger dropdown value.
+            switchMap(() => {
+              return this.store
+                .select(BuilderSelectors.selectCurrentStep)
+                .pipe(take(1));
+            }),
+            switchMap((res) => {
+              if (res?.type === TriggerType.PIECE) {
+                return this.actionMetaService
+                  .getPieceMetadata(
+                    res.settings.pieceName,
+                    res.settings.pieceVersion
+                  )
+                  .pipe(
+                    map((meta) => {
+                      return { step: res, metadata: meta };
+                    })
+                  );
+              }
+              return of({ step: res, metadata: undefined });
+            }),
+            tap(() => {
+              this.store.dispatch(FlowsActions.toggleWaitingToSave());
+            }),
+            debounceTime(350),
+            tap((res) => {
+              console.log(res.step?.displayName);
+              if (
```

Comment:
I feel this can be simplified by adding switch cases

Reply:
will fix this and the as in another pr, let's release